### DISCUSSION TITLE
US915 Join Frequency Pattern Randomisation

### DISCUSSION THREAD

**Discussion by Carl-P4** [Lora-net/LoRaMac-node#1178] [Q&A] [answered]
Using Tag V4.4.4 Via ST Supplied LRWAN_V2.0.0 package,
We have noticed that the first join at power-up of the device is always in the same band, the first 8 channels of the first band.

How do you direct the first join to a randomised band, so that a number of the same devices powered up at the same time don't attempt to use the same band.

Re-joins after the first join appear to be randomised and occur using different channels it seems, which is what is needed.

Also we are using ADR, should the application manually vary the Datarate the join is requested at, again to reduce air time if we are close the gateway.

![image](https://user-images.githubusercontent.com/9988707/136058624-3bdb916a-ab65-4bce-90de-a5ae06b86f7a.png)

.. EDIT ..
After further investigation I thought LmHandlerJoin needed to be called again if a join fails. But debug logging has shown the stack will automatically cycle through different frequencies attempting to join.

Testing on a single unit shows the sequence of frequencies it tries to always be the same order. Tested version V4.4.7. US915

0s069:TX on freq 902500000 Hz at DR 0
10s126:TX on freq 905300000 Hz at DR 0
20s126:TX on freq 905500000 Hz at DR 0
30s126:TX on freq 908100000 Hz at DR 0
40s126:TX on freq 909700000 Hz at DR 0
50s126:TX on freq 910900000 Hz at DR 0
60s126:TX on freq 912500000 Hz at DR 0
70s126:TX on freq 913700000 Hz at DR 0

How can I force these join attempts to be more random so that if multiple units are powered up their join sequences are all different.

It looks like Nvm.RegionGroup1.JoinTrialsCounter increments for each failed attempt. But I dont see an interface to adjust it.

What is the suggested method for randomising this process and if I want to adjust the join attempt rate, how is this done?

Thanks,

**Comment by mluis1** [comment, marked as answer]
We don't know which changes ST has made on their fork of this project.

By default this project uses all 64+1 channels. The join frequencies selection follows the recommendation provided by the [Regional Parameters specification](https://lora-alliance.org/wp-content/uploads/2020/11/rp_2-1.0.1.pdf) 

![image](https://user-images.githubusercontent.com/5763295/137292987-781c3bb7-9915-4eae-aad0-b5bdabaf4c39.png)

The channels frequency selection order should be always random as the MAC layer uses a random number generated by the Radio hardware to seed the pseudo random generator. This means that every time `LoRaMacInitialization` API is called a different pseudo random sequence will be used.

https://github.com/Lora-net/LoRaMac-node/blob/05353ff70bec5349a26b23a02168accb0ea03ad9/src/mac/LoRaMac.c#L3835-L3836

Therefore, the only possibility to always get the same sequence across end-devices would be if the Radio always return the same random number.
The only possibility for the radio to return the same random number can only happen if the Rx RF path is saturated by an high power RF signal close to the end-device (almost unlikely to happen).

**Reply by Carl-P4** [reply]
Hrm that is odd then, It seemed quite repeatable.

Can you point me at where the stack selects the next frequency to attempt a join on? I can then look at how random the selection is

**Reply by mluis1** [reply]
The channel selection for US915 band is done by the following function

https://github.com/Lora-net/LoRaMac-node/blob/05353ff70bec5349a26b23a02168accb0ea03ad9/src/mac/region/RegionUS915.c#L803-L891

And

https://github.com/Lora-net/LoRaMac-node/blob/05353ff70bec5349a26b23a02168accb0ea03ad9/src/mac/region/RegionBaseUS.c#L75-L135

**Reply by mluis1** [reply]
Something that you can also verify is if the `srand1` function is called only 1 time. The call from `LoRaMacInitialization`.

**Reply by Carl-P4** [reply]
Will do, shall report back if I find anything. Although the version I'm using from ST isn't your latest

**Reply by mluis1** [reply]
The functions for version 4.4.4 are
https://github.com/Lora-net/LoRaMac-node/blob/ba17382bd5109513937afad07f068a781a503ef6/src/mac/region/RegionUS915.c#L956-L1045

and

https://github.com/Lora-net/LoRaMac-node/blob/ba17382bd5109513937afad07f068a781a503ef6/src/mac/region/RegionUS915.c#L141-L196

**Reply by Carl-P4** [reply]
The issue appears to be with the random seed initialisation, breakpoint shows the seed is consistently set to 0. 
I'm using a Murata LoRa module with ST V2.1.0 (LoRaWAN_End_Node example unaltered) which uses V4.4.7.
SX1276Random function in sx1276.c

This runs a loop asking the radio for the RSSI.
SX1276Read returns 0 for every call

Radio interface works as I can join and transmit etc..

**Reply by mluis1** [reply]
This explains why you always observe the same pseudo random sequence.

Now we need to understand why the SX1276Read always returns a 0.
The possible causes are:
- The RF receive path is saturated by a close by strong signal. 
   Could you try to put the device inside a shielded box?
   Like this we could confirm that there is no external influence causing the issue.
- The radio is not in receiver mode for an yet unknown reason.
   Could you try to add a 1 ms delay after the call to `SX1276SetOpMode( RF_OPMODE_RECEIVER );` on `SX1276Random` function?

**Reply by mluis1** [reply]
I have just tried with a `NucleoL476+SX1276MB1LAS` platform and random numbers are being generated.
1st trial:
![image](https://user-images.githubusercontent.com/5763295/137444876-906cadd3-15fc-4890-8d67-364fd7024f82.png)

2nd trial:
![image](https://user-images.githubusercontent.com/5763295/137445029-2a5fac8a-8530-4350-8f49-45f6e53c3c99.png)

**Reply by Carl-P4** [reply]
Yeah I used break points to slow down the execution when I saw the return values in the loop were all 0. In case it was a timing issue on the SX1276SetOpMode command. Stepping over the loop to observe all 0's

I don't believe it's RF saturation as I can communicate with the gateway well.

Just for fun I added a delay and a call to read the OP Mode, after setting it to RF_OPMODE_RECEIVER and it returns 0.

![image](https://user-images.githubusercontent.com/9988707/137449485-a6638c49-b5fe-4f82-8606-eb41730a4030.png)

Did you test with the latest build or V4.4.7? 

At this point I am assuming this to be either an old issue you've solved or an ST induced one.
I can see the stack call SX1276Init before the call to SX1276Random

**Reply by mluis1** [reply]
I have downloaded the latest [I-CUBE-LRWAN](https://www.st.com/content/my_st_com/en/products/embedded-software/mcu-mpu-embedded-software/stm32-embedded-software/stm32cube-expansion-packages/i-cube-lrwan.license=1634287311039.product=I-CUBE-LRWAN.version=2.1.0.html) (v2.1.0) package from ST web site and the random numbers are also being generated by the SX1276 radio.

I have used again the `NucleoL476+SX1276MB1LAS`.
I launched `LoRaWAN_End_Node` Keil project under `en.i-cube_lrwan_v2.1.0\STM32CubeExpansion_LRWAN_V2.1.0\Projects\NUCLEO-L476RG\Applications\LoRaWAN\LoRaWAN_End_Node`. Then I selected `sx1276mb1las` target and I enabled the debug pins under `sys_conf.h`, compiled and launched a debug session.
Please see below screenshots for a couple of results.

I would double check all the modifications you have done on the ST provided example code.
I'm out of ideas on how to help you debug your issue.

1st trial:
![image](https://user-images.githubusercontent.com/5763295/137462490-45209b26-53a9-4de8-98a5-483415c0a42a.png)

2nd trial:
![image](https://user-images.githubusercontent.com/5763295/137462716-97626d0a-6a7e-45b7-825d-a37ee22fdad2.png)

**Reply by Carl-P4** [reply]
Interesting, I don't think i've changed much more than that, other than enabling both EU868 and US915 regions.  But will restart from the supplied ST Zip to confirm

**Reply by mluis1** [reply]
On my side it is also working if I enable EU868 and US915 region. I have also tried using EU868 and US915 as active regions.

Do you have the possibility to try with another hardware? 
Just in case your current hardware is malfunctioning for some reason.

**Reply by Carl-P4** [reply]
Ahh I think I've seen a difference. When I moved to the latest ST build for the tests, I didn't change TCXO Enable pin assignment, which would cause the radio not to function and then not provide a RSSI for the seed.

**Reply by mluis1** [reply]
Indeed that could be the reason for the observed issue.

**Reply by Carl-P4** [reply]
Another thing I've noticed is the RegionBaseUSComputeNext125kHzJoinChannel chooses a band to try based upon groupsCurrentIndex which is supplied from the NVM structure (RegionNvmGroup1->JoinChannelGroupsCurrentIndex). If you are not using NVM this always starts at zero putting the first attempt in the first band? Is having the app randomise this value on boot helpful.

**Reply by mluis1** [reply]
Changing the way  `RegionNvmGroup1->JoinChannelGroupsCurrentIndex` is initialized will not bring any improvement as the randomly selected sub-band will not necessarily be the one you are expecting and the JoinRequests may cycle anyway up until you reach the right sub-band.

The US915 and similar regions are expected to run under 64 channels gateways and everything would work smoothly. However, a lot of LoRaWAN networks are only deploying 8 channels gateways. Doing so, introduces drawbacks and everyone must accept the compromises that come with it. The compromise is that one needs to accept that it may take several trials up until we reach the right sub-band. Once the network is joined then everything will work as expected as the Network Server will tell the end-devices what channels it can use.
